Force specialization on INT_LIST

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111219
* __->__ #111216

Follow up on https://github.com/pytorch/pytorch/pull/95479

Fixes https://github.com/pytorch/pytorch/issues/111198

Fixes https://github.com/pytorch/pytorch/issues/111197

Fixes https://github.com/pytorch/pytorch/issues/111188

Fixes https://github.com/pytorch/pytorch/issues/111201

Fixes https://github.com/pytorch/pytorch/issues/111202

I can also do this for some other types, will do this stacked on top.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang committed:
Force specialization on INT_LIST

Follow up on https://github.com/pytorch/pytorch/pull/95479

Fixes https://github.com/pytorch/pytorch/issues/111198

I can also do this for some other types, will do this stacked on top.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

jon-chuang commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -658,6 +658,9 @@ inline std::vector<int64_t> PythonArgs::intlistWithDefault(
         }
+      } else if (torch::is_symint(py::handle(obj))) {
+        res[idx] = py::cast<c10::SymInt>(py::handle(obj))
+                       .guard_int(__FILE__, __LINE__);
```
quite curious, why don't we guard in symintlist case?

https://github.com/pytorch/pytorch/blob/8162f4170b8c7df3bab503103544e1d7cbf8b052/torch/csrc/utils/python_arg_parser.h#L534

jon-chuang replied:
Is this for the so-called unbacked symint case? :thinking:

jon-chuang replied:
Oh right, this guard is necessary to extract the backed values, which the downstream aten op requires?

ezyang replied:
You don't need to guard for symintlist because we can just return the SymInts directly, to track the symbolic flow. That's the point of having SymInts.

The guard here is for operators that don't advertise symint support. Stuff like dim=[1, 2]. We never handle this symbolically, always specialize.

jon-chuang replied:
So essentially, they flow through the program as a DAG, and can terminate at leaves where they are specialized. I suppose there are also scenarios where they are outputs of the program and are never specialized.

ezyang replied:
They form a DAG when they're not specialized. If we specialize, we don't actually need to model its compute because we just inline constants everywhere.

ezyang committed:
Update on "Force specialization on INT_LIST"


Follow up on https://github.com/pytorch/pytorch/pull/95479

Fixes https://github.com/pytorch/pytorch/issues/111198

Fixes https://github.com/pytorch/pytorch/issues/111197

Fixes https://github.com/pytorch/pytorch/issues/111188

I can also do this for some other types, will do this stacked on top.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang committed:
Update on "Force specialization on INT_LIST"


Follow up on https://github.com/pytorch/pytorch/pull/95479

Fixes https://github.com/pytorch/pytorch/issues/111198

Fixes https://github.com/pytorch/pytorch/issues/111197

Fixes https://github.com/pytorch/pytorch/issues/111188

I can also do this for some other types, will do this stacked on top.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -658,6 +658,9 @@ inline std::vector<int64_t> PythonArgs::intlistWithDefault(
         }
+      } else if (torch::is_symint(py::handle(obj))) {
+        res[idx] = py::cast<c10::SymInt>(py::handle(obj))
+                       .guard_int(__FILE__, __LINE__);
```
More for my own understanding: instead of silently specializing whenever we plumb a symint to an op who's schema specifies `int[]`, wouldn't it be better to just change all of our schemas to take in `SymInt[]`? (and until then, raising an error makes it easier to identify and fix the places where this happens?)

One reason the guards here are kinda painful is that (I don't think) we get cpp stacktraces in the guard provenance tracking, so it won't be immediately clear to people who sees guards that the guards are coming from our unfixed schemas.

bdhirsh replied:
oh unless this change is purely for BC reasons - in theory we could change all aten schemas today, but I guess for custom ops, we used to specialize by default whereas now (without this PR) we now error by default on custom op schemas that take in `int[]`.

ezyang replied:
So what we specifically are guarding against here is when Dynamo thinks an int is dynamic (most commonly because dynamic=True) but it actually is something like dim= and cannot possibly be dynamic.

I think modifying the schemas would not be the right thing to do, because there isn't any situation we will actually support dynamism for real, so it is wasted work. In the early days, we errored instead of specializing to make it easier to catch missing symints in schema, but we now have good logging and good coverage so it is more important to avoid erroring eg as in the linked issues

ezyang committed:
Update on "Force specialization on INT_LIST"


Follow up on https://github.com/pytorch/pytorch/pull/95479

Fixes https://github.com/pytorch/pytorch/issues/111198

Fixes https://github.com/pytorch/pytorch/issues/111197

Fixes https://github.com/pytorch/pytorch/issues/111188

I can also do this for some other types, will do this stacked on top.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
